File: sdks/cpp/connections/gRPC/src/controllers/Connect.cpp
/*
 * Copyright 2024 Ross Video Ltd
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its
 * contributors may be used to endorse or promote products derived from this
 * software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS “AS IS”
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * RE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

// connections/gRPC
#include <controllers/Connect.h>
#include <Logger.h>
using catena::gRPC::Connect;
using catena::common::ILanguagePack;

// Initializes the object counter for Connect to 0.
int catena::gRPC::Connect::objectCounter_ = 0;

/**
 * Constructor which initializes and registers the current Connect object, 
 * then starts the process.
 */
catena::gRPC::Connect::Connect(ICatenaServiceImpl *service, SlotMap& dms, bool ok)
    : CallData(service), writer_(&context_),
        status_{ok ? CallStatus::kCreate : CallStatus::kFinish}, 
        catena::common::Connect(dms, service->getSubscriptionManager()) {
    service_->registerItem(this);
    objectId_ = objectCounter_++;
    proceed(ok);  // start the process
}

// Manages gRPC command execution process using the state variable status.
void catena::gRPC::Connect::proceed(bool ok) {
    DEBUG_LOG << "Connect proceed[" << objectId_ << "]: " << timeNow()
                << " status: " << static_cast<int>(status_) << ", ok: "
                << std::boolalpha << ok;

    /**
     * The newest connect object (the one that has not yet been attached to a
     * client request) will send shutdown signal to cancel all open connections
     */
    if (!ok && status_ != CallStatus::kFinish) {
        DEBUG_LOG << "Connect[" << objectId_ << "] cancelled";
        DEBUG_LOG << "Cancelling all open connections";
        shutdownSignal_.emit();
        status_ = CallStatus::kFinish;
    }

    std::unique_lock<std::mutex> lock{mtx_, std::defer_lock};
    switch (status_) {
        /** 
         * kCreate: Updates status to kProcess and requests the Connect command
         * from the service.
         */ 
        case CallStatus::kCreate:
            status_ = CallStatus::kProcess;
            service_->RequestConnect(&context_, &req_, &writer_, service_->cq(), service_->cq(), this);
            break;
        /**
         * kProcess: Processes the request asyncronously, updating status to
         * kFinish and notifying the responder once finished.
         */
        case CallStatus::kProcess:
            // Used to serve other clients while processing.
            new Connect(service_, dms_, ok);
            context_.AsyncNotifyWhenDone(this);
            try {
                catena::exception_with_status rc{"", catena::StatusCode::OK};
                initAuthz_(jwsToken_(), service_->authorizationEnabled());
                // Cancels all open connections if shutdown signal is sent.
                shutdownSignalId_ = shutdownSignal_.connect([this](){
                    context_.TryCancel();
                    hasUpdate_ = true;
                    this->cv_.notify_one();
                });

                // Set detail level from request
                detailLevel_ = req_.detail_level();

                // Connecting to each device in dms_.
                for (auto [slot, dm] : dms_) {
                    if (dm) {
                        // Waiting for a value set by server to be sent to execute code.
                        valueSetByServerIds_[slot] = dm->getValueSetByServer().connect([this, slot](const std::string& oid, const IParam* p){
                            updateResponse_(oid, p, slot);
                        });
                        // Waiting for a value set by client to be sent to execute code.
                        valueSetByClientIds_[slot] = dm->getValueSetByClient().connect([this, slot](const std::string& oid, const IParam* p){
                            updateResponse_(oid, p, slot);
                        });
                        // Waiting for a language to be added to execute code.
                        languageAddedIds_[slot] = dm->getLanguageAddedPushUpdate().connect([this, slot](const ILanguagePack* l) {
                            updateResponse_(l, slot);
                        });
                        // Send client a empty update with slot of the device
                        catena::PushUpdates populatedSlots;
                        populatedSlots.set_slot(slot);
                        writer_.Write(populatedSlots, this);
                    }
                }

                status_ = CallStatus::kWrite;

            } catch (catena::exception_with_status& rc) {
                status_ = CallStatus::kFinish;
                grpc::Status errorStatus(static_cast<grpc::StatusCode>(rc.status), rc.what());
                writer_.Finish(errorStatus, this);
            }
            break;
        /**
         * kWrite: Waits until an update to either set res to device's slot or
         * end the process.
         */
        case CallStatus::kWrite:
            lock.lock();
            cv_.wait(lock, [this] { return hasUpdate_; });
            hasUpdate_ = false;
            // If connect was cancelled finish the process.
            if (context_.IsCancelled()) {
                status_ = CallStatus::kFinish;
                DEBUG_LOG << "Connection[" << objectId_ << "] cancelled";
                writer_.Finish(Status::CANCELLED, this);
                break;
            // Send the client an update with the slot of the device.
            } else {
                writer_.Write(res_, this);
            }
            lock.unlock();
            break;
        /**
         * kFinish: Ends the connection.
         */
        case CallStatus::kFinish:
            DEBUG_LOG << "Connect[" << objectId_ << "] finished";
            // Disconnecting all initialized listeners.
            if (shutdownSignalId_ != 0) { shutdownSignal_.disconnect(shutdownSignalId_); }
            for (auto [slot, dm] : dms_) {
                if (dm) {
                    if (valueSetByClientIds_.contains(slot)) {
                        dm->getValueSetByClient().disconnect(valueSetByClientIds_[slot]);
                    }
                    if (valueSetByServerIds_.contains(slot)) {
                        dm->getValueSetByServer().disconnect(valueSetByServerIds_[slot]);
                    }
                    if (languageAddedIds_.contains(slot)) {
                        dm->getLanguageAddedPushUpdate().disconnect(languageAddedIds_[slot]);
                    }
                }
            }
            service_->deregisterItem(this);
            break;
        // default: Error, end process.
        default:
            status_ = CallStatus::kFinish;
            grpc::Status errorStatus(grpc::StatusCode::INTERNAL, "illegal state");
            writer_.Finish(errorStatus, this);
    }
}

// Returns true if the connection has been cancelled.
bool catena::gRPC::Connect::isCancelled() {
    return context_.IsCancelled();
}

File: sdks/go/pkg/rest/connect.go
/*
 * Copyright 2026 Ross Video Ltd
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its
 * contributors may be used to endorse or promote products derived from this
 * software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

package rest

import (
	"encoding/json"
	"fmt"
	"net/http"

	"github.com/rossvideo/catena/sdks/go/pkg/catena"
	"github.com/rossvideo/catena/sdks/go/pkg/logger"
)

// PushUpdate represents a streaming update sent to connected clients via SSE
type PushUpdate struct {
	Slot            int    `json:"slot,omitempty"`
	OID             string `json:"oid,omitempty"`
	Value           any    `json:"value,omitempty"`
	SlotsAdded      []int  `json:"slots_added,omitempty"`
	SlotsRemoved    []int  `json:"slots_removed,omitempty"`
	DeviceComponent any    `json:"device_component,omitempty"`
}

// sseConnection represents an active SSE connection
type sseConnection struct {
	updates chan PushUpdate
	slots   []int // slots this connection is subscribed to
}

// SetMaxConnections sets the maximum number of SSE connections allowed (0 = unlimited)
func (s *Server) SetMaxConnections(max int) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.maxConnections = max
}

// registerConnection registers a new SSE connection and returns its ID
func (s *Server) registerConnection(slots []int) (int, chan PushUpdate) {
	s.mu.Lock()
	defer s.mu.Unlock()

	// Check connection limit
	if s.maxConnections > 0 && len(s.connections) >= s.maxConnections {
		return -1, nil
	}

	s.nextConnID++
	updates := make(chan PushUpdate, 100)
	s.connections[s.nextConnID] = &sseConnection{
		updates: updates,
		slots:   slots,
	}
	logger.Info("SSE connection registered", "connID", s.nextConnID, "total", len(s.connections))
	return s.nextConnID, updates
}

// unregisterConnection removes an SSE connection
func (s *Server) unregisterConnection(connID int) {
	s.mu.Lock()
	defer s.mu.Unlock()

	if conn, ok := s.connections[connID]; ok {
		close(conn.updates)
		delete(s.connections, connID)
		logger.Info("SSE connection unregistered", "connID", connID, "remaining", len(s.connections))
	}
}

// BroadcastUpdate sends an update to all connected SSE clients
func (s *Server) BroadcastUpdate(update PushUpdate) {
	s.mu.Lock()
	defer s.mu.Unlock()

	for connID, conn := range s.connections {
		// Only send to connections subscribed to this slot (or all if slot is 0)
		if update.Slot == 0 || len(conn.slots) == 0 || containsSlot(conn.slots, update.Slot) {
			select {
			case conn.updates <- update:
				// Successfully sent
			default:
				// Channel full, log warning but don't block
				logger.Warning("SSE channel full, dropping update", "connID", connID, "oid", update.OID)
			}
		}
	}
}

// sendSSEEvent writes a single SSE event to the response writer
func (s *Server) sendSSEEvent(w http.ResponseWriter, flusher http.Flusher, update PushUpdate) error {
	data, err := json.Marshal(update)
	if err != nil {
		return err
	}
	_, err = fmt.Fprintf(w, "data: %s\n\n", data)
	if err != nil {
		return err
	}
	flusher.Flush()
	return nil
}

// getPopulatedSlots returns the list of registered device slots
func (s *Server) getPopulatedSlots() []int {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.slots
}

// containsSlot checks if a slot is in the list
func containsSlot(slots []int, slot int) bool {
	for _, s := range slots {
		if s == slot {
			return true
		}
	}
	return false
}

// handleConnect handles GET /st2138-api/v1/connect (SSE streaming)
func (s *Server) handleConnect(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		val, res := catena.ReplyError[catena.CatenaValue](catena.METHOD_NOT_ALLOWED, "only GET allowed")
		writeHTTPResult(w, res, val)
		return
	}
	// Check if SSE streaming is supported
	flusher, ok := w.(http.Flusher)
	if !ok {
		val, res := catena.ReplyError[catena.CatenaValue](catena.INTERNAL, "streaming not supported")
		writeHTTPResult(w, res, val)
		return
	}

	// Register this connection
	connID, updates := s.registerConnection(s.getPopulatedSlots())
	if connID < 0 {
		val, res := catena.ReplyError[catena.CatenaValue](catena.RESOURCE_EXHAUSTED, "too many connections")
		writeHTTPResult(w, res, val)
		return
	}
	defer s.unregisterConnection(connID)

	// Set SSE headers
	w.Header().Set("Content-Type", "text/event-stream")
	w.Header().Set("Cache-Control", "no-cache")
	w.Header().Set("Connection", "keep-alive")
	origin := r.Header.Get("Origin")
	if origin != "" {
		w.Header().Set("Access-Control-Allow-Origin", origin)
		w.Header().Set("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE, OPTIONS")
		w.Header().Set("Access-Control-Allow-Headers", "Content-Type, Authorization, Accept, Origin, X-Requested-With, Language, Detail-Level")
		w.Header().Set("Access-Control-Allow-Credentials", "true")
	}
	w.WriteHeader(http.StatusOK)
	flusher.Flush()

	// Send initial message with populated slots
	initialUpdate := PushUpdate{
		SlotsAdded: s.getPopulatedSlots(),
	}
	if err := s.sendSSEEvent(w, flusher, initialUpdate); err != nil {
		logger.Error("failed to send initial SSE event", "error", err)
		return
	}

	logger.Info("SSE Connect started", "connID", connID)

	// Stream updates until client disconnects
	ctx := r.Context()
	for {
		select {
		case <-ctx.Done():
			// Client disconnected
			logger.Info("SSE client disconnected", "connID", connID)
			return
		case update, ok := <-updates:
			if !ok {
				// Channel closed (server shutdown or forced disconnect)
				logger.Info("SSE channel closed", "connID", connID)
				return
			}
			if err := s.sendSSEEvent(w, flusher, update); err != nil {
				logger.Error("failed to send SSE event", "connID", connID, "error", err)
				return
			}
		}
	}
}
